test/proxy/ERC1967/ERC1967Utils.test.js

```diff
@@ -0,0 +1,181 @@
+} = require('../../helpers/erc1967');
+
+const { ZERO_ADDRESS } = constants;
+
+const ERC1967Utils = artifacts.require('$ERC1967Utils');
+
+const V1 = artifacts.require('DummyImplementation');
+const V2 = artifacts.require('DummyImplementationV2');
+const UpgradeableBeaconMock = artifacts.require('UpgradeableBeaconMock');
+
+contract('ERC1967Utils', function (accounts) {
+  const [, admin, anotherAccount] = accounts;
+  const EMPTY_DATA = '0x';
+
+  beforeEach('setup', async function () {
+    this.utils = await ERC1967Utils.new();
+    this.v1 = await V1.new();
+    this.v2 = await V2.new();
+  });
+
+  describe('IMPLEMENTATION_SLOT', function () {
+    beforeEach('set v1 implementation', async function () {
+      await setSlot(this.utils, ImplementationSlot, this.v1.address);
+    });
+
+    describe('getImplementation', function () {
+      it('returns current implementation and matches implementation slot value', async function () {
+        expect(await this.utils.$getImplementation()).to.equal(this.v1.address);
+        expect(await getAddressInSlot(this.utils.address, ImplementationSlot)).to.equal(this.v1.address);
+      });
+    });
+
+    describe('upgradeToAndCall', function () {
+      it('sets implementation in storage and emits event', async function () {
+        const newImplementation = this.v2.address;
+        const receipt = await this.utils.$upgradeToAndCall(newImplementation, EMPTY_DATA);
+
+        expect(await getAddressInSlot(this.utils.address, ImplementationSlot)).to.equal(newImplementation);
+        expectEvent(receipt, 'Upgraded', { implementation: newImplementation });
+      });
+
+      it('reverts when implementation does not contain code', async function () {
+        await expectRevertCustomError(
+          this.utils.$upgradeToAndCall(anotherAccount, EMPTY_DATA),
+          'ERC1967InvalidImplementation',
+          [anotherAccount],
+        );
+      });
+
+      describe('when data is empty', function () {
+        it('reverts when value is sent', async function () {
+          await expectRevertCustomError(
+            this.utils.$upgradeToAndCall(this.v2.address, EMPTY_DATA, { value: 1 }),
+            'ERC1967NonPayable',
+            [],
+          );
+        });
+      });
+
+      describe('when data is not empty', function () {
+        it('delegates a call to the new implementation', async function () {
+          const valueSlot = 0;
+          const newVal = web3.utils.hexToNumber('0xdeabeef');
+
+          // Initialize by calling migrate and modifying the slot 0 `value`
+          const initializeData = this.v2.contract.methods.migrate(newVal).encodeABI();
+          await this.utils.$upgradeToAndCall(this.v2.address, initializeData);
+
+          // Because of the delegatecall to `migrate`, the ERC1967Utils will hold the new value at slot 0
+          const slotValue = await getSlot(this.utils.address, valueSlot);
+          expect(web3.utils.hexToNumber(slotValue)).to.equal(newVal);
```

Comment:
This test and the other similar one are failing in the upgradeable checks, because the transpiler adds `Initializable` with its own storage and this pushes the storage layout. We shouldn't make assumptions about storage positions in tests.

It might be better to test this by with an initialization function that emits an event.